### PR TITLE
Fix some issues found by PVS-Studio #1

### PR DESCRIPTION
**Pull request by dimhotepus**
Product page: https://www.viva64.com/en/pvs-studio/
How to use for free: https://www.viva64.com/en/b/0600/

### CODE REVIEW

**Comment by WonderCooper** [comment]
您好，已收到您的邮件，我会尽快处理，谢谢

**Comment by gongminmin** [comment]
First, if the PR is a commercial of PVS-Studio, it's not welcomed. Second, some of the changes are cool, but some are not necessary. I can't separate them if they are all in one PR.

**Comment by dimhotepus** [comment]
> First, if the PR is a commercial of PVS-Studio, it's not welcomed. Second, some of the changes are cool, but some are not necessary. I can't separate them if they are all in one PR.

PR is not commercial, just a demonstration of product possibilities. PVS-Studio is free for OSS.
> The free license cannot be applied to the following types of projects:
>
> 1. commercial (the author profits from the project directly or indirectly);
> 2. owned by organizations (the project has a corporate profile on GitHub and/or other ownership types);
> 3. developed by organizations (financing and/or hiring developers);
> 4. mirrors and forks (for example, copies of projects that do not meet these conditions).
>
> The PVS-Studio license for Open Source is intended for personal creative projects. This license allows the author to accept donations from their community.

Regarding second issue, could you please mentions which commits are not required and I'll drop them?

File: KlayGE/Core/Include/KlayGE/RenderSettings.hpp [hunk 1]
```diff
@@ -34,9 +34,9 @@ namespace KlayGE
 	{
 		RenderSettings()
 			: hide_win(false), full_screen(false),
-				left(0), top(0),
+				left(0), top(0), width(0), height(0),
```
Comment (gongminmin) [inline KlayGE/Core/Include/KlayGE/RenderSettings.hpp]: These changes are unnecessary. They are guaranteed to be assigned before the first usage.

File: KlayGE/Core/Include/KlayGE/UI.hpp [hunk 2]
```diff
@@ -140,7 +140,7 @@ namespace KlayGE
 					is_mouse_over_(false), has_focus_(false), is_default_(false),
 					x_(0), y_(0), width_(0), height_(0),
 					dialog_(dialog), index_(0),
-					id_(0), type_(type), enabled_(true),
+					id_(0), type_(type), hotkey_(0), enabled_(true),
```
Comment (gongminmin) [inline KlayGE/Core/Include/KlayGE/UI.hpp]: Same as this one. Unnecessary change.

File: KlayGE/Core/Src/Base/HWDetect.cpp [hunk 3]
```diff
@@ -108,9 +108,10 @@ namespace KlayGE
 		{
 			BOOST_ASSERT(wbem_services_);
 
-			wchar_t query_lang[] = L"WQL";
+			BSTR query_lang = SysAllocString(L"WQL");
```
Comment (gongminmin) [inline KlayGE/Core/Src/Base/HWDetect.cpp:111]: BSTR is wchar_t* on WIndows. No need to change this file.

File: KlayGE/Core/Src/Input/Touch.cpp [hunk 4]
```diff
@@ -49,7 +49,7 @@ namespace KlayGE
 		: wheel_delta_(0), index_(false), num_available_touch_(0),
 			gesture_(TS_None), action_param_(MakeSharedPtr<InputTouchActionParam>()),
 			curr_state_(GS_None),
-			one_finger_tap_timer_(0), two_finger_tap_timer_(0)
+			one_finger_tap_timer_(0), two_finger_tap_timer_(0), two_finger_start_len_(0)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Input/Touch.cpp]: No need to change.

File: KlayGE/Core/Src/Render/CameraController.cpp [hunk 5]
```diff
@@ -457,7 +457,7 @@ namespace KlayGE
 
 
 	CameraPathController::CameraPathController()
-		: curr_frame_(0), frame_rate_(30)
+		: start_time_(0), curr_frame_(0), frame_rate_(30)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/CameraController.cpp]: No need to change.

File: KlayGE/Core/Src/Render/DistanceField.cpp [hunk 6]
```diff
@@ -192,8 +192,8 @@ namespace
 
 	void ComputeGradient(std::vector<float> const& img, int w, int h, std::vector<float2>& grad)
 	{
-		BOOST_ASSERT(img.size() == static_cast<size_t>(w * h));
-		BOOST_ASSERT(grad.size() == static_cast<size_t>(w * h));
+		BOOST_ASSERT(img.size() == static_cast<size_t>(w) * static_cast<size_t>(h));
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/DistanceField.cpp]: No need to change these w * h. The are far less than causing overflow.

File: KlayGE/Core/Src/Render/FrameBuffer.cpp [hunk 7]
```diff
@@ -26,7 +26,8 @@ namespace KlayGE
 {
 	FrameBuffer::FrameBuffer()
 					: left_(0), top_(0), width_(0), height_(0),
-						viewport_(MakeSharedPtr<KlayGE::Viewport>())
+						viewport_(MakeSharedPtr<KlayGE::Viewport>()),
+						views_dirty_(false)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/FrameBuffer.cpp]: No need to change.

File: KlayGE/Core/Src/Render/IndirectLightingLayer.cpp [hunk 8]
```diff
@@ -57,6 +57,7 @@ namespace KlayGE
 	MultiResSILLayer::MultiResSILLayer()
 	{
 		multi_res_layer_ = MakeUniquePtr<MultiResLayer>();
+		g_buffer_camera_ = nullptr;
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/IndirectLightingLayer.cpp]: No need to change.

File: KlayGE/Core/Src/Render/JudaTexture.cpp [hunk 9]
```diff
@@ -120,7 +120,10 @@ namespace KlayGE
 		: root_(MakeSharedPtr<QuadTreeNode>()),
 			num_tiles_(num_tiles), tile_size_(tile_size), format_(format),
 			texel_size_(NumFormatBytes(format)),
-			decode_tick_(0), tile_tick_(0)
+			data_blocks_offset_(0),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/JudaTexture.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Light.cpp [hunk 10]
```diff
@@ -631,6 +631,7 @@ namespace KlayGE
 	SphereAreaLightSource::SphereAreaLightSource()
 	{
 		type_ = LT_SphereArea;
+		radius_ = 0;
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Light.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Mesh.cpp [hunk 11]
```diff
@@ -537,6 +537,7 @@ namespace KlayGE
 
 	StaticMesh::StaticMesh(std::wstring_view name)
 		: Renderable(name),
+			mtl_id_(0),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Mesh.cpp]: No need to change.

File: KlayGE/Core/Src/Render/PostProcess.cpp [hunk 12]
```diff
@@ -246,7 +246,8 @@ namespace KlayGE
 			: Renderable(name),
 				volumetric_(volumetric),
 				cs_based_(false), cs_pixel_per_thread_x_(1), cs_pixel_per_thread_y_(1), cs_pixel_per_thread_z_(1),
-				num_bind_output_(0)
+				num_bind_output_(0), pp_mvp_param_(nullptr),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/PostProcess.cpp]: No need to change.

File: KlayGE/Core/Src/Render/RenderLayout.cpp [hunk 13]
```diff
@@ -30,6 +30,7 @@ namespace KlayGE
 				start_index_location_(0),
 				base_vertex_location_(0),
 				start_instance_location_(0),
+				indirect_args_offset(0),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/RenderLayout.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Texture.cpp [hunk 14]
```diff
@@ -1628,15 +1628,15 @@ namespace
 namespace KlayGE
 {
 	Texture::Mapper::Mapper(Texture& tex, uint32_t array_index, uint32_t level, TextureMapAccess tma, uint32_t x_offset, uint32_t width)
-		: tex_(tex), mapped_array_index_(array_index), mapped_level_(level)
+		: tex_(tex), mapped_array_index_(array_index), mapped_face_(CF_Positive_X), mapped_level_(level)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Texture.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Texture.cpp [hunk 15]
```diff
@@ -1628,15 +1628,15 @@ namespace
...
 	{
 		tex_.Map1D(array_index, level, tma, x_offset, width, data_);
 		row_pitch_ = slice_pitch_ = width * NumFormatBytes(tex.Format());
 	}
 
 	Texture::Mapper::Mapper(Texture& tex, uint32_t array_index, uint32_t level, TextureMapAccess tma, uint32_t x_offset, uint32_t y_offset,
 		uint32_t width, uint32_t height)
-		: tex_(tex), mapped_array_index_(array_index), mapped_level_(level)
+		: tex_(tex), mapped_array_index_(array_index), mapped_face_(CF_Positive_X), mapped_level_(level)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Texture.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Texture.cpp [hunk 16]
```diff
@@ -1646,7 +1646,7 @@ namespace KlayGE
 
 	Texture::Mapper::Mapper(Texture& tex, uint32_t array_index, uint32_t level, TextureMapAccess tma, uint32_t x_offset, uint32_t y_offset,
 		uint32_t z_offset, uint32_t width, uint32_t height, uint32_t depth)
-		: tex_(tex), mapped_array_index_(array_index), mapped_level_(level)
+		: tex_(tex), mapped_array_index_(array_index), mapped_face_(CF_Positive_X), mapped_level_(level)
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Texture.cpp]: No need to change.

File: KlayGE/Core/Src/Render/Texture.cpp [hunk 17]
```diff
@@ -2262,7 +2262,8 @@ namespace KlayGE
 							base[index] = data_block.size();
 							data_block.resize(base[index] + init_data[index].slice_pitch * the_depth);
 
-							tex_res->read(&data_block[base[index]], static_cast<std::streamsize>(init_data[index].slice_pitch * the_depth));
+							tex_res->read(&data_block[base[index]],
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Texture.cpp]: No need to change. Impossible to overflow.

File: KlayGE/Core/Src/Render/Texture.cpp [hunk 18]
```diff
@@ -3122,7 +3123,8 @@ namespace KlayGE
 
 
 	Texture::Texture(Texture::TextureType type, uint32_t sample_count, uint32_t sample_quality, uint32_t access_hint)
-		: type_(type), sample_count_(sample_count), sample_quality_(sample_quality), access_hint_(access_hint)
+		: num_mip_maps_(0), array_size_(0), format_(EF_Unknown),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Render/Texture.cpp]: No need to change.

File: KlayGE/Core/Src/Scene/SceneManager.cpp [hunk 19]
```diff
@@ -68,6 +68,7 @@ namespace KlayGE
 			overlay_root_(L"OverlayRoot", SceneNode::SOA_Cullable | SceneNode::SOA_Overlay),
 			small_obj_threshold_(0),
 			update_elapse_(1.0f / 60),
+			urt_(0),
```
Comment (gongminmin) [inline KlayGE/Core/Src/Scene/SceneManager.cpp]: No need to change.

File: KlayGE/Core/Src/UI/UIPolylineEditBox.cpp [hunk 20]
```diff
@@ -27,6 +27,7 @@ namespace KlayGE
 	UIPolylineEditBox::UIPolylineEditBox(uint32_t type, UIDialogPtr const & dialog)
 					: UIControl(type, dialog),
 						active_pt_(-1),
+						move_over_pt_(-1),
```
Comment (gongminmin) [inline KlayGE/Core/Src/UI/UIPolylineEditBox.cpp]: No need to change.

**Review by gongminmin — CHANGES_REQUESTED** [review]

**Comment by gongminmin** [comment]
I've made a round of code review. Basically, no need to change the member initialization, BSTR, or overflow. They are not problems.

File: KlayGE/Core/Src/Base/HWDetect.cpp (hunk 3 above)
Comment (dimhotepus) [reply KlayGE/Core/Src/Base/HWDetect.cpp:111]: Unfortunately it is not correct. Read https://learn.microsoft.com/en-us/previous-versions/windows/desktop/automat/bstr:

A BSTR is a composite data type that consists of a length prefix, a data string, and a terminator. The following table describes these components.

The following code is incorrect:
 
> BSTR MyBstr = L"I am a happy BSTR";

This code builds (compiles and links) correctly, but it will not function properly because the string does not have a length prefix. If you use a debugger to examine the memory location of this variable, you will not see a four-byte length prefix preceding the data string.

Instead, use the following code:

> BSTR MyBstr = SysAllocString(L"I am a happy BSTR");

A debugger that examines the memory location of this variable will now reveal a length prefix containing the value 34. This is the expected value for a 17-byte character string that is represented as a wide-string literal through the inclusion of the "L" string prefix. The debugger will also show the terminator that appears after the data string.

**If you pass a wide-string literal as an argument to a COM function that is expecting a BSTR, the COM function behaves unexpectedly.**

Comment (gongminmin) [reply KlayGE/Core/Src/Base/HWDetect.cpp:111]: However, in many examples from Microsoft, wide strings are directly used as BSTR. E.g, https://learn.microsoft.com/en-us/windows/win32/api/wbemcli/nn-wbemcli-iwbemservices?source=recommendations, https://learn.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemlocator-connectserver. In the implementation, they actually accept strings with or without length prefix. It doesn't matter at least when using with IWbemServices.

Comment (dimhotepus) [reply KlayGE/Core/Src/Base/HWDetect.cpp:111]: Oh, even their docs are inconsistent and contain errors. I would prefer to use recommended way and create MR to fix incorrect usage in docs you have mentioned.
Anyway, you decide which way should we use.

Comment (dimhotepus) [reply KlayGE/Core/Src/Base/HWDetect.cpp:111]: I've landed 2 PRs to update Microsoft docs:
https://github.com/MicrosoftDocs/sdk-api/pull/1511
https://github.com/MicrosoftDocs/sdk-api/pull/1512

Now BSTR and wchar_t[] will not be intermixed and samples provide correct way of WMI API usage

**Review by gongminmin — APPROVED** [review]
Thanks for your contribution.

**Comment by WonderCooper** [comment]
您好，已收到您的邮件，我会尽快处理，谢谢
